src/userdb/userdb.8.in
.\"  <!-- $Id: userdb.8.in,v 1.1.1.1 2003/05/07 02:13:35 lfan Exp $ -->
.\"  <!-- Copyright 1998 - 2001 Double Precision, Inc.  See COPYING for -->
.\"  <!-- distribution information. -->
.\" This manpage has been automatically generated by docbook2man 
.\" from a DocBook document.  This tool can be found at:
.\" <http://shell.ipoline.com/~elmert/comp/docbook2X/> 
.\" Please send any bug reports, improvements, comments, patches, 
.\" etc. to Steve Cheng <steve@ggi-project.org>.
.TH "USERDB" "8" "12 December 2002" "Double Precision, Inc." ""
.SH NAME
userdb \- manipulate @userdb@
.SH SYNOPSIS

\fBuserdb\fR \fB\fIaddr\fB\fR \fBset\fR \fB\fIfield\fB=\fIvalue\fB\fR\fI ...\fR


\fBuserdb\fR \fB\fIaddr\fB\fR \fBunset\fR \fB\fIfield\fB\fR\fI ...\fR


\fBuserdb\fR \fB\fIaddr\fB\fR \fBdel\fR


\fBuserdb\fR \fB\fIpath/addr\fB\fR [ \fBset\fR | \fBunset\fR | \fBdel\fR ] \fB...\fR


\fBuserdb\fR \fB-f\fR \fB\fIfile\fB\fR \fB\fIadr\fB\fR [ \fBset\fR | \fBunset\fR | \fBdel\fR ] \fB...\fR


\fBuserdb\fR \fB-show\fR \fB\fIpath\fB\fR


\fBuserdb\fR \fB-show\fR \fB\fIpath\fB\fR \fB\fIaddr\fB\fR


\fBuserdb\fR \fB-show\fR \fB-f\fR \fB\fIfile\fB\fR


\fBuserdb\fR \fB-show\fR \fB-f\fR \fB\fIfile\fB\fR \fB\fIaddr\fB\fR

.SH "DESCRIPTION"
.PP
\fBuserdb\fR is a convenient script to individually manipulate
entries in \fI@userdb@\fR. See
\fBmakeuserdb\fR(8)
for a description of its contents.  \fI@userdb@\fR can always
be edited using any text editor, but \fBuserdb\fR is a
convenient way to modify this file from another script.
.PP
\fI@userdb@\fR can also be a subdirectory, instead of a file.
Specify \fB\fIfoo/bar/addr\fB\fR to manipulate
\fB\fIaddr\fB\fR in the file
\fI@userdb@/foo/bar\fR.  You can
also use the
\fB-f\fR flag: \fB-f
\fI@userdb@/foo/bar\fB\fR is equivalent.  Use
whatever form makes the most sense to you.
.PP
\fI@userdb@\fR must not have any group or world
permissions. That's
because its contents may include system passwords (depending upon the
application which uses this virtual user account database).
.PP
Each line in \fI@userdb@\fR takes following form:
.sp
.RS
.PP
\fIaddr\fR<TAB>\fIfield\fR=\fIvalue\fR|\fIfield\fR=\fIvalue\fR...
.RE
.PP
\fIaddr\fR specifies a unique virtual address. It
is followed by a single
tab character, then a list of
\fIfield\fR=\fIvalue\fR pairs,
separated by
vertical slash characters. See
\fBmakeuserdb\fR(8)
for field definitions.
.PP
A text editor can be used to add blank lines or comments in
\fI@userdb@\fR.  Any blank lines or comments are ignored by the
\fBuserdb\fR script.
.PP
The names of the actual fields, and their contents, are defined entirely by
applications that use the \fI@userdb@\fR database, the
\fBuserdb\fR command just adds or removes arbitrary fields.
.PP
For example:
.sp
.RS
.PP

.nf
\fBuserdb default/info set mail=/home/mail/info\fR
.fi
.RE
.PP
This command accesses the address "info" in
\fI@userdb@/default\fR.
.PP
If the second argument to \fBuserdb\fR is
"\fIset\fR", the
remaining arguments are taken as
\fIfield=value\fR pairs, which are
added to the record for \fIaddr\fR. If there is no
record for \fIaddr\fR, a
new record will be appended to the file. If
\fIaddr\fR exists, any existing
values of any specified fields are removed. If
\fI=value\fR is missing,
\fBuserdb\fR stops and prompts for it. This is useful if
you're setting
a password field, where you do not want to specify the password on the command
line, which can be seen by the
\fBps\fR(1)
command. If \fBuserdb\fR is being
executed by a script, the value can be provided on standard input.
.PP
Use "\fIunset\fR" to delete fields from an existing
record. Use
"\fIdel\fR" to delete all fields in the existing record,
plus the record itself.
.SS "DISPLAYING @USERDB@"
.PP
If the first argument to userdb
is \fI-show\fR, \fBuserdb\fR
displays the contents of \fI@userdb@\fR. If
\fI@userdb@\fR is a
subdirectory,
\fIpath\fR must refer to a
specific file in \fI@userdb@\fR. The
\fI-f\fR option can be used instead of
\fIpath\fR in order to specify an
arbitrary file.
.PP
If
\fIaddr\fR is not specified,
\fBuserdb\fR produces a list, on standard
output, containing all addresses found in the file, on per line. If
\fIaddr\fR is specified,
\fBuserdb\fR produces a list, on standard output, of
all the fields in \fI@userdb@\fR for this
\fIaddr\fR.
.SS "REBUILDING @USERDB@.DAT"
.PP
The actual virtual account/address database is
\fI@userdb@.dat\fR.
This is a binary database file. \fB@userdb@\fR is the plain text
version. After running \fBuserdb\fR, execute the
\fBmakeuserdb\fR(8) command to rebuild
\fI@userdb@.dat\fR for the changes to take effect.
.SH "BUGS"
.PP
\fIaddr\fR must be unique.
If \fI@userdb@\fR is a subdirectory,
it's possible to create the same
\fIaddr\fR
in different files in the subdirectory.
This is an error that is not currently detected by \fBuserdb\fR,
however the subsequent
\fBmakeuserdb\fR(8) command
will fail with an error message.
.SH "FILES"
.PP
\fI @userdb@\fR - plain text file, or directory of plain text files
.PP
\fI .lock.filename\fR - lock file for \fIfilename\fR
.PP
\fI .tmp.filename\fR - temporary file used to create new contents of \fIfilename\fR
.SH "SEE ALSO"
.PP
\fBmakeuserdb\fR(8),
\fBuserdbpw\fR(8)


src/userdb/userdb.c
/*
** Copyright 1998 - 2001 Double Precision, Inc.
** See COPYING for distribution information.
*/

#if	HAVE_CONFIG_H
#include	"config.h"
#endif
#include	"dbobj.h"
#include	"userdb.h"
#include	<string.h>
#include	<stdlib.h>
#include	<errno.h>
#include	<time.h>
#include	<sys/types.h>
#if	HAVE_SYS_STAT_H
#include	<sys/stat.h>
#endif

static const char rcsid[]="$Id: userdb.c,v 1.1.1.1 2003/05/07 02:13:34 lfan Exp $";

static struct dbobj d;
static time_t dt;
static ino_t di;

static int initialized=0;

/* Open userdb.dat, if already opened, see if it changed, if so reopen */

void userdb_init(const char *n)
{
struct	stat	stat_buf;

	if (initialized)
	{
		if (stat(n, &stat_buf) ||
			stat_buf.st_mtime != dt ||
			stat_buf.st_ino != di)
		{
			dbobj_close(&d);
			initialized=0;
			dt=stat_buf.st_mtime;
			di=stat_buf.st_ino;
		}
	}
	else if (stat(n, &stat_buf))
	{
		return;
	}
	else
	{
		dt=stat_buf.st_mtime;
		di=stat_buf.st_ino;
	}

	if (!initialized)
	{
		if (dbobj_open(&d, n, "R"))
		{
			return;
		}
		initialized=1;
	}
}

void userdb_close()
{
	if (initialized)
	{
		dbobj_close(&d);
		initialized=0;
	}
}

/* Fetch a record from userdb.dat */

char	*userdb(const char *u)
{
char	*p,*q;
size_t	l;

	if (!initialized)
	{
		errno=ENOENT;
		return (0);
	}

	q=dbobj_fetch(&d, u, strlen(u), &l, "");
	if (!q)
	{
		errno=ENOENT;
		return(0);
	}

	p=malloc(l+1);
	if (!p)	return (0);

	if (l)	memcpy(p, q, l);
	free(q);
	p[l]=0;
	return (p);
}

/* Return a pointer to a specific field in this record */

const char	*userdb_get(const char *u, const char *n, int *l)
{
int	nl=strlen(n);

	while (u && *u)
	{
		if (memcmp(u, n, nl) == 0 &&
			(u[nl] == 0 || u[nl] == '=' || u[nl] == '|'))
		{
			u += nl;
			*l=0;
			if (*u == '=')
			{
				++u;
				while ( u[*l] && u[*l] != '|')
					++ *l;
			}
			return (u);
		}
		u=strchr(u, '|');
		if (u)	++u;
	}
	return (0);
}

/* Extract field as an unsigned int */

unsigned userdb_getu(const char *u, const char *n, unsigned defnum)
{
	int l;
	const char *p;

	if ((p=userdb_get(u, n, &l)) != 0)
	{
		defnum=0;
		while (l && *p >= '0' && *p <= '9')
		{
			defnum = defnum * 10 + (*p++ - '0');
			--l;
		}
	}
	return (defnum);
}

/* Extract a field into a dynamically allocated buffer */

char *userdb_gets(const char *u, const char *n)
{
	int l;
	const char *p;
	char	*q;

	if ((p=userdb_get(u, n, &l)) != 0)
	{
		q=malloc(l+1);
		if (!q)
			return (0);

		if (l)	memcpy(q, p, l);
		q[l]=0;
		return (q);
	}
	errno=ENOENT;
	return (0);
}

/* Create a userdbs structure based upon a uid (reverse lookup) */

struct userdbs *userdb_createsuid(uid_t u)
{
char	buf[80];
char	*p=buf+sizeof(buf)-1, *q;
struct	userdbs *s;

	/* Lookup uid= record */

	*p=0;
	*--p='=';
	do
	{
		*--p= "0123456789"[u % 10];
		u=u/10;
	} while (u);
	p=userdb(p);
	if (!p)	return (0);

	/* Have account name, now look it up. */

	q=userdb(p);
	if (!q)
	{
		free(p);
		return (0);
	}
	s=userdb_creates(q);
	if (s)
		s->udb_name=p;
	else
		free(p);
	free(q);
	return (s);
}

/* Extracted a userdb.dat record, convert it to a userdbs structure */

struct userdbs *userdb_creates(const char *u)
{
struct userdbs *udbs=(struct userdbs *)malloc(sizeof(struct userdbs));
char	*s;

	if (!udbs)	return (0);
	memset((char *)udbs, 0, sizeof(*udbs));

	if ((udbs->udb_dir=userdb_gets(u, "home")) == 0)
	{
		userdb_frees(udbs);
		return (0);
	}

	if ((s=userdb_gets(u, "uid")) != 0)
	{
		udbs->udb_uid=atol(s);
		free(s);
		if ((s=userdb_gets(u, "gid")) != 0)
		{
			udbs->udb_gid=atol(s);
			free(s);

			if ((s=userdb_gets(u, "shell")) != 0)
				udbs->udb_shell=s;
			else if (errno != ENOENT)
			{
				userdb_frees(udbs);
				return (0);
			}

			if ((s=userdb_gets(u, "mail")) != 0)
				udbs->udb_mailbox=s;
			else if (errno != ENOENT)
			{
				userdb_frees(udbs);
				return (0);
			}
			if ((s=userdb_gets(u, "quota")) != 0)
				udbs->udb_quota=s;
			else if (errno != ENOENT)
			{
				userdb_frees(udbs);
				return (0);
			}
			if ((s=userdb_gets(u, "gecos")) != 0)
				udbs->udb_gecos=s;
			else if (errno != ENOENT)
			{
				userdb_frees(udbs);
				return (0);
			}
			udbs->udb_source=userdb_gets(u, "_");
			return (udbs);
		}
	}
	userdb_frees(udbs);
	return (0);
}

void	userdb_frees(struct userdbs *u)
{
	if (u->udb_name)	free(u->udb_name);
	if (u->udb_gecos)	free(u->udb_gecos);
	if (u->udb_dir)		free(u->udb_dir);
	if (u->udb_shell)	free(u->udb_shell);
	if (u->udb_mailbox)	free(u->udb_mailbox);
	if (u->udb_quota)	free(u->udb_quota);
	if (u->udb_source)	free(u->udb_source);
	free(u);
}

